crates/valence_registry/src/tags.rs

```diff
@@ -0,0 +1,95 @@
+use bevy_ecs::prelude::*;
+use bytes::Bytes;
+use serde::Deserialize;
+use valence_core::ident::Ident;
+use valence_core::protocol::encode::encode_packet;
+use valence_core::protocol::packet::synchronize_tags::*;
+
+#[derive(Debug, Resource)]
+pub struct TagsRegistry {
+    pub registries: Vec<Registry>,
+    cached_packet: Bytes,
+}
+
+#[derive(Debug, Deserialize)]
+pub struct Registry {
+    #[serde(rename = "registry")]
+    pub kind: Ident<String>,
+    pub tags: Vec<TagEntry>,
+}
+
+#[derive(Debug, Deserialize)]
+pub struct TagEntry {
+    pub name: Ident<String>,
+    #[serde(rename = "values")]
+    pub entries: Vec<u64>,
+}
+
+impl TagsRegistry {
+    pub(crate) fn build_synchronize_tags(&self) -> SynchronizeTagsS2c<'_> {
+        SynchronizeTagsS2c {
+            tags: self
+                .registries
+                .iter()
+                .map(|registry| TagGroup {
+                    kind: registry.kind.as_str_ident().into(),
+                    tags: registry
+                        .tags
+                        .iter()
+                        .map(|tag| Tag {
+                            name: tag.name.as_str_ident().into(),
+                            entries: tag.entries.clone(),
+                        })
+                        .collect(),
+                })
+                .collect(),
+        }
+    }
+
+    pub fn sync_tags_packet(&self) -> &Bytes {
+        &self.cached_packet
+    }
+}
+
+pub fn init_tags_registry(mut tags: ResMut<TagsRegistry>) {
+    let registries =
+        serde_json::from_str::<Vec<Registry>>(include_str!("../../../extracted/tags.json"))
+            .expect("tags.json is invalid");
+    tags.registries = registries;
+}
+
+pub fn cache_tags_packet(tags: ResMut<TagsRegistry>) {
+    if tags.is_changed() {
+        let tags = tags.into_inner();
+        let packet = tags.build_synchronize_tags();
+        let mut bytes = vec![];
+        encode_packet(&mut bytes, &packet).expect("failed to encode tags packet");
```

Comment:
Ah, this needs to use a temporary `PacketWriter` with `.write_packet` instead of using `encode_packet`. You'll have to get the global compression threshold from `Res<Server>`. (sorry it's a bit confusing. `encode_packet` should probably be made private.)